Доверительные IP адреса / gRPC

rnikbond opened the pull request.

rnikbond committed:
Сервер: добавлена обработка списка доверительных IP адресов, добавлены тесты для middleware.
Клиент: Добавлен заголовок X-Real-IP при отправке запроса.

rnikbond committed:
Убран отладочный вывод

rnikbond committed:
Убран дефолтный список доверительных IP адресов на сервере

rnikbond committed:
Добавлен gRPC сервер

rnikbond committed:
Добавлена отправка метрик через gRPC

rnikbond committed:
Добавлен хеш метрики в сообщение grpc.
Добавлена остановка grpc сервера.

Barugoo commented on cmd/server/main.go:
```diff
@@ -79,16 +79,30 @@ func main() {
 
+	gServ, err := server.NewGRPCServer(cfg.AddrRPC, storeManager)
+	if err != nil {
+		logger.Err.Printf("failed create gRPC server^ %v\n", err)
```
А есть ли нам смысл сделать делать `else`? Я бы просто здесь фаталил и `gserv.Start()` вызывал вне ветвления:
```go
gServ, err := server.NewGRPCServer(cfg.AddrRPC, storeManager)
 if err != nil {
 	logger.Fatal.Printf("failed create gRPC server^ %v\n", err)
}
gServ.Start()

Barugoo replied:
Кажется, что если у нас был передан конфиг для gRPC-сервера, то он либо должен стартовать, либо фаталится. Иначе у тебя стартанет приложение с предположительно работающим gRPC-сервером, контейнер будет работать и вроде как все ОК но о том, что у тебя gRPC ручки не работают ты узнаешь только когда сделаешь к ним запрос. При rolling-выкатке это может привести к плохой ситуации, тк код с точки зрения оркестратора рабочий, а gRPC-клиенты тем временем не могут достучаться и мы получаем даунтайм.
Общее правило для таких вещей - если у тебя законфигурированы какие-то режимы работы (это касается и используемых хранилок, и используемых контроллеров, и каких-то иных вещей) приложения, то сервис должен целиком работать с учетом всех конфигураций, либо не стартовать вовсе.

Barugoo commented on cmd/server/main.go:
```diff
@@ -79,16 +79,30 @@ func main() {
 	<-ctx.Done()
 	stop()
 
+	if gServ != nil {
```
Ну и эта проверка собсна тоже не нужна, если у тебя AddrGRPC пустой. Если он не пустой (то есть gRPC сервер ДОЛЖЕН работать), то исходя из логики моего предыдущего комментария при неработающем gRPC сервере у тебя процесс упадет десятью строчками раннее

Barugoo commented on internal/agent/services/reporter/reporter.go:
```diff
@@ -166,6 +152,69 @@ func (r Reporter) Report(ctx context.Context, reportType string) error {
+// reportGRPC Отправка метрик GRPC шлюз
+func (r Reporter) reportGRPC(ctx context.Context) error {
+
+	parts := strings.Split(r.addr, ":")
+	if len(parts) == 0 {
+		return fmt.Errorf("invalid address grpc gate")
+	}
+
+	conn, err := grpc.Dial(":"+parts[len(parts)-1], grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if errClose := conn.Close(); errClose != nil {
+			r.logger.Err.Printf("failed close gPRC connection: %v\n", errClose)
+		}
+	}()
+
+	c := pb.NewMetricsClient(conn)
```
А зачем мы создаем новое подключение на каждой итерации репорта? Давай один раз при старте приложения где-нибудь в функции `main` инициализируем gRPC клиент и пробросим его полем в структуру `Reporter`

Barugoo commented on proto/metrics.proto:
```diff
@@ -0,0 +1,26 @@
+}
+
+message UpsertResponse {
+  string error = 1;
```
Кажется, что нам не нужна в теле респонса ошибка, в gRPC мы ее возвращаем в любом случае 
Я бы просто в сигнатуре методов сделал `returns (google.protobuf.Empty)` если нам нечего возвращать

Barugoo commented on internal/server/grpc_server.go:
```diff
@@ -0,0 +1,89 @@
+
+	var response pb.UpsertResponse
+
+	metric, err := metricPkg.CreateMetric(
+		metricPkg.GaugeType,
+		in.Id,
+		metricPkg.WithValueFloat(in.Value),
+	)
+
+	if err == nil {
+		err = serv.m.Upsert(metric)
+	}
+
+	if err != nil {
+		response.Error = err.Error()
+	}
+
+	return &response, nil
```
Соответственно, если мы убираем ошибку из тела `UpsertResponse` (вместе с самой структурой `UpsertResponse`, тк больше полей в ней не остается), то здесь нам надо вернуть `&emptypb.Empty{}`  
Также избавимся от множества телодвижений с переменной `err`, из-за которой на мой взгляд трудновато читать код. 
Получится что-то такое:
```go
import "google.golang.org/protobuf/types/known/emptypb"
//...
	metric, err := metricPkg.CreateMetric(
 		metricPkg.GaugeType,
 		in.Id,
 		metricPkg.WithValueFloat(in.Value),
 	)
 	if err != nil {
 		return nil, status.Errorf(codes.InternalServerError, "unable to create metric entity: %v", err)
 	}

        if err := serv.m.Upsert(metric); err != nil {
 		return nil, status.Errorf(codes.InternalServerError, "unable to save metric: %v", err)
 	}
 	return &emptypb.Empty{}, nil
}
```

Barugoo reviewed (changes requested):
Хорошо, но есть замечания

rnikbond committed:
Теперь сервер не запускается, если gRPC сервер не запустился из-за ошибки конфигурации

rnikbond committed:
Proto: Свое сообщение Response заменено на google.pb.Empty.
При ошибке обновления метрики на сервере через gRPC, информация об ошибке записывается в возвращаемую ошибку (раньше записывалась в Response).

rnikbond committed:
Агнет: создание gRPC соединения вынесено в создание агента, теперь соединение создается один раз.

Barugoo commented on internal/server/grpc_server.go:
```diff
@@ -0,0 +1,80 @@
+	)
+
+	if err != nil {
+		return &emptypb.Empty{}, err
+	}
+
+	return &emptypb.Empty{}, serv.m.Upsert(metric)
```
Кстати, чтобы не инициализировать каждый раз, можно просто в начале функции сделать:
```go
res := &emptypb.Empty{}
///...
	metric, err := metricPkg.CreateMetric(
 		metricPkg.GaugeType,
 		in.Id,
 		metricPkg.WithValueFloat(in.Value),
 	)

 	if err != nil {
 		return res, err
 	}
```

или же в сигнатуре сделать именованые возвращаемые значения (не опуская при этом `return` в теле):
```go
func (serv *MetricsServiceRPC) UpsertGauge(ctx context.Context, in *pb.UpsertGaugeRequest) (res *emptypb.Empty, err error) {
//...
	metric, err := metricPkg.CreateMetric(
 		metricPkg.GaugeType,
 		in.Id,
 		metricPkg.WithValueFloat(in.Value),
 	)

 	if err != nil {
 		return res, err
 	}

rnikbond committed:
Рефакторинг
